### PR TITLE
Advanced Querying for ChainReader

### PR DESCRIPTION
**Pull request by silaslenihan**
Corresponding ticket: [BCFR-44](https://smartcontract-it.atlassian.net/jira/software/c/projects/BCFR/boards/358?assignee=61a6f090ef18ca00717f84db&selectedIssue=BCFR-909)

### Requires

https://github.com/smartcontractkit/chainlink-common/pull/798
https://github.com/smartcontractkit/chainlink-solana/pull/869

### Resolves

[BCFR-44]: https://smartcontract-it.atlassian.net/browse/BCFR-44?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/services/relay/evm/chain_reader.go [hunk 1]
```diff
@@ -456,6 +445,52 @@ func getEventTypes(event abi.Event) ([]abi.Argument, types.CodecEntry, map[strin
 	return indexedAsUnIndexedTypes, types.NewCodecEntry(indexedTypes, nil, nil), dataWords
 }
 
+func findFieldIndex(arg abi.Argument, fieldPath string, dataWords map[string]read.DataWordDetail, index int) int {
```
Comment (ilija42) [inline core/services/relay/evm/chain_reader.go]: wdyt about, and add a short comment
```suggestion
func calculateFieldDWIndex(arg abi.Argument, fieldPath string, dataWords map[string]read.DataWordDetail, index int) int {
```

**Review by ilija42 — COMMENTED** [review]
Great job! Left a small nit comment, but otherwise looks good

File: core/services/relay/evm/chain_reader.go (hunk 1 above)
Comment (silaslenihan) [reply core/services/relay/evm/chain_reader.go]: refactored field name and added a comment above

**Review by ilija42 — DISMISSED** [review]

File: core/services/relay/evm/read/event.go [hunk 2]
```diff
@@ -548,6 +548,12 @@ func (b *EventBinding) encodeComparator(comparator *primitives.Comparator) (quer
 }
 
 func (b *EventBinding) encodeValComparatorDataWord(dwTypeID string, value any) (hash common.Hash, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("%w: cannot encode %s data word comparator", commontypes.ErrInvalidType, dwTypeID)
```
Comment (jmank88) [inline core/services/relay/evm/read/event.go]: Why are we expecting a panic?

Comment (ilija42) [reply core/services/relay/evm/read/event.go]: I added this, geth abi package in some scenarios panics instead of throwing an error

Comment (jmank88) [reply core/services/relay/evm/read/event.go]: How about mentioning "recovered from panic" somewhere in the returned error? And `r` is critical to include as well.

Comment (silaslenihan) [reply core/services/relay/evm/read/event.go]: Updated the error message accordingly.

**Review by DavidOrchard — APPROVED** [review]

**Review by ilija42 — APPROVED** [review]

**Pull request merged**